auacm/app/modules/competition_manager/views.py

```diff
@@ -181,3 +202,149 @@ def get_competition_data(cid):
+        ))
+        socketio.emit('new_user', {
+            'cid': cid,
+            'user': {
+                'display': user.display,
+                'username': user.username
+            }
+        },
+        namespace='/register')
+    session.flush()
+    session.commit()
+
+    return serve_response({})
+
+
+@app.route('/api/competitions/<int:cid>/unregister', methods=['POST'])
+@login_required
+def unregister_for_competition(cid):
+    """ Called when a user wants to register for a competition.
+
+    All the user has to do is submit a post to this url with no form data.
+    From their logged-in status, we'll go ahead and remove them from the
+    competiton.
+
+    Similar to the <code>/register</code> endpoint, an admin can post a list of
+    users to unregister from the competition.
+    """
+    if session.query(Competition).filter(Competition.cid == cid).first() \
+            is None:
+        return serve_error('Competition does not exist', response_code=404)
+
+    if current_user.admin == 1 and 'users' in request.data:
+        registrants = loads(request.data['users'])
+    else:
+        registrants = list()
+        registrants.append(current_user.username)
+
+    for user in registrants:
+        (session.query(CompUser)
+                .filter(CompUser.username == user, CompUser.cid == cid)
+                .delete())
+    session.flush()
+    session.commit()
+
+    return serve_response({})
+
+
+@app.route('/api/competitions/<int:cid>/teams', methods=['GET'])
+@login_required
+def get_competition_teams(cid):
+    """ Get all of the teams in a competition.
+
+    Returns all of the teams, their users, and those users' display names.
+    """
+    comp_users = session.query(CompUser, User).join(User,
+            User.username == CompUser.username).filter(CompUser.cid == cid)\
+            .all()
+
+    teams = dict()
+    for user in comp_users:
+        if user.CompUser.team not in teams:
+            teams[user.CompUser.team] = list()
+        teams[user.CompUser.team].append({
+            'username': user.User.username,
+            'display': user.User.display
+        })
+
+    return serve_response(teams)
+
+
+@app.route('/api/competitions/<int:cid>/teams', methods=['PUT'])
+@login_required
+def put_competition_teams(cid):
+    """ Update the teams for a competition
+
+    If a user is an admin, they can update the competition's users, doing a PUT.
+    This will take the JSON data in the 'teams' part of the request form and
+    store it to the database. Any teams or users not included int the JSON data
+    will not be a part of the competition and will have to re-register; however
+    it should not be used for the solely purpose of de-registering participants.
+    """
+    if current_user.admin == 0:
+        return serve_error('You must be an admin to update teams',
+                response_code=401)
+
+    try:
+        teams = loads(request.form['teams'])
+    except KeyError as err:
```

Comment:
It may be a good idea to catch a ValueError if the JSON isn't properly formed. Kinda falls under form validation